Issue#1099: Add a custom text block

docwilmot opened the pull request:
Fixes https://github.com/backdrop/backdrop-issues/issues/1099
Replaces https://github.com/backdrop/backdrop/pull/1092

quicksketch commented on core/modules/layout/includes/custom_text_block.inc:
```diff
@@ -0,0 +1,89 @@
+      $form_state = array();
+      form_process_fieldset($fieldset, $form_state);
+      $preview .= backdrop_render($fieldset);
+    }
```
This looks good, but it's a lot of copy/paste from the Block::getAdminPreview() function. Perhaps we should split it up so that child-blocks can inherit or call a parent function to reduce duplication.

What if we had something like this:

```
function getAdminPreview() {
   $label_text = !empty($this->settings['subject']) ? ': ' . check_plain($this->settings['subject']) : '';
   $preview = '<p>Custom text' . $label_text;
   $preview .= '</p>';
   $preview .= $this->getAdminConditionsPreview();
```

Where getAdminConditionsPreview() was a new protected function in the parent `Block` class?

quicksketch commented on core/modules/layout/includes/custom_text_block.inc:
```diff
@@ -0,0 +1,89 @@
+    if (!empty($this->settings['content'])) {
+      return check_markup(token_replace($this->settings['content']), $this->settings['format'], '', TRUE);
+    }
+    return 'No content saved';
```
We should translate "No content saved". But is this going out to the front-end of the user's site? I'm not sure we should special-case anything at all here. Or maybe we only show this message to users that have "administer layouts" permission so that they can actually do something about it.

docwilmot replied:
I've removed it. If there is no subject or content the block won't render right?

quicksketch replied:
> I've removed it. If there is no subject or content the block won't render right?

Yes. And actually the entire block might disappear from view (including the wrappers). I'd need to double-check.

docwilmot commented:
I'm having a doubt about how to handle the title. Currently I have it so the config 'title' field is the block title and also used as the label in the Layout draggable block; if you leave the 'title' field blank, the block won't have a title, and the draggable block wont have a label. So we don't use the standard title select found in most block dialogs with 'Default', 'Custom' and 'None'. 

Should we have this though so you can have a block subject, but hide it using the select? So the config title field will be used as the block title and the label if 'Default', and just the label if its 'Hidden' or 'Custom'?

quicksketch commented:
> Currently I have it so the config 'title' field is the block title and also used as the label in the Layout draggable block; if you leave the 'title' field blank, the block won't have a title, and the draggable block wont have a label.

Seems like we should just have an Admin Label in addition to the Title then? Something by which it could always be identified, even if the block does not have a visible title on the front-end.

docwilmot commented:
Actually the Layout draggable will always have a minimum label (just "Custom text", not an identifiable label). I was thinking along these lines:

```
function form(&$form, &$form_state) {
  $token_link = ' ' . theme('token_tree_link');
  $form['title_display']['title_display'] = array(
    '#type' => 'select',
    '#title' => t('Block title'),
    '#options' => array(
      LAYOUT_TITLE_DEFAULT => t('Use subject'),
      LAYOUT_TITLE_NONE => t('None'),
    ),
    '#default_value' => $this->settings['title_display'],
  );
  $form['subject'] = array(
    '#type' => 'textfield',
    '#title' => t('Subject'),
    '#default_value' => !empty($this->settings['subject']) ? $this->settings['subject'] : '',
  );
  $form['content'] = array(
    '#type' => 'text_format',
    '#title' => t('Block content'),
    '#default_value' => !empty($this->settings['content']) ? $this->settings['content'] : '',
    '#description' => $token_link,
    '#token_types' => array(),
  );
}
```

So if you wanted a subject (and draggable label) to ID your blocks you can type in a subject but choose 'None" in the select if you want no block title on the front end. I think this is the best of both worlds.

We could have a `LAYOUT_TITLE_CUSTOM => t('Custom'),` key on the `title_display` as well. So a person could have a separate title (shown on block view) from block subject (shown on the draggable), which would satisfy the idea of an Admin label as you suggest.

jenlampton commented:
This PR is looking really good. The only feedback I have is on the name. I think "Custom block" and "Custom text block" are too similar and will be confusing. What's wrong with calling this a "Text block"? or just "Text" ?

jenlampton commented on core/modules/layout/includes/custom_text_block.inc:
```diff
@@ -0,0 +1,89 @@
+   */
+  function getAdminPreview() {
+    $label_text = !empty($this->settings['subject']) ? ': ' . check_plain($this->settings['subject']) : '';
+    $preview = '<p>Custom text' . $label_text;
```
why split this onto two lines? `$preview = '<p>Custom text' . $label_text . '</p>';` would be better, but we should translate `Custom text` if we decide to leave it here.

Also, why put the title in the preview instead of the content? Shouldn't we put the title into the title preview, and content into the content?

jenlampton commented on core/modules/layout/includes/custom_text_block.inc:
```diff
@@ -0,0 +1,89 @@
+    $label_text = !empty($this->settings['subject']) ? ': ' . check_plain($this->settings['subject']) : '';
+    $preview = '<p>Custom text' . $label_text;
+    $preview .= '</p>';
+    
```
whitespace here should be removed.

jenlampton commented:
also it looks like a file is being added to core at `core/modules/core/modules/layout/includes/custom_text_block.inc` we should double check that path.

edit: nevermind. I think that was me applying the patch incorrectly.

docwilmot commented:
Updated:
- Tests added, passing
- Changed to just plain ole Text Block so as not to confuse with Custom blocks
- Updated Labels
- Removed Token Browser as token insertion doesn't yet work with our CKEditor

quicksketch commented on core/modules/layout/includes/block.text.inc:
```diff
@@ -0,0 +1,69 @@
+   *  Sets title text on draggable block panel in Layout builder.
+   */
+  function getAdminTitle() {
+    return t('Text block');
```
Let's move the display label up into the admin title (`$this->settings['subject']`). For reusable blocks (once we have those) we can use the admin label (`info`) instead.

quicksketch commented on core/modules/layout/includes/block.text.inc:
```diff
@@ -0,0 +1,69 @@
+  function getAdminPreview() {
+    $label_text = !empty($this->settings['subject']) ? ': ' . check_plain($this->settings['subject']) : '';
+    $preview = '<p>' . t('Text') . $label_text . '</p>';
+    $preview .= $this->getAdminConditionsPreview();
```
The description here can be "A custom text block." Or you could even remove the entire thing I think and just let the `Block::getAdminPreview()` handle it.

quicksketch commented on core/modules/layout/layout.module:
```diff
@@ -601,6 +601,7 @@ function layout_autoload_info() {
     'Block' => 'includes/block.class.inc',
     'BlockBroken' => 'includes/block.class.inc',
     'BlockLegacy' => 'includes/block_legacy.class.inc',
+    'TextBlock' => 'includes/block.text.inc',
```
I feel like we should keep this under the `Block` namespace (not 100% on that) and name this `BlockText` instead of `TextBlock`. TextBlock would be a block owned by text.module. Of course, Layout module is hijacking Block's namespaces here, but I think that's okay.

quicksketch commented on core/modules/layout/layout.module:
```diff
@@ -1533,3 +1534,17 @@ function layout_set_handler_contexts(array &$contexts, array $all_contexts, arra
+  $blocks['text_block'] = array(
+    'info' => t('Text block'),
+    'description' => t('Create a text block.'),
+    'class' => 'TextBlock',
```
Conferring with @jenlampton, terminology here could be changed:
- info: "Add custom block"
- description: "A basic block for adding custom text."
- class: BlockText

This way the dialog will show "Add custom block" (as you will be able to do once we integrate this with block.module), and the description "A basic block for adding custom text." will be shown in the admin preview.

quicksketch commented on core/modules/layout/tests/layout.test:
```diff
@@ -243,6 +243,69 @@ class LayoutInterfaceTest extends BackdropWebTestCase {
   /**
+   * Test Custom Text blocks.
+   */
+  function testCustomTextBlocks() {
```
Yay, very nice!

quicksketch commented on core/modules/layout/includes/block.class.inc:
```diff
@@ -136,7 +136,16 @@ class Block extends LayoutHandler {
+  /**
+   * Return a fieldset listing any visibility conditions set for this block.
+   */
+  function getAdminConditionsPreview() {
```
This currently conflicts with the PR we made for one field per block. I like your function name `getAdminConditionsPreview` better though, let's keep that instead of `getConditionSummary()`, which is what we have currently.

docwilmot commented:
Tests fail because we're changing the link to add one of these from "Add text block" to "Add a custom block".

quicksketch commented:
> Tests fail because we're changing the link to add one of these from "Add text block" to "Add a custom block".

Ah, of course. Yay tests. :wink:

quicksketch commented:
I've fixed the tests and filed a new PR at https://github.com/backdrop/backdrop/issues/1198.
